code/modules/ai/ai_behaviors/carbon/xeno/xeno.dm

```diff
@@ -44,34 +44,35 @@
 			change_state(REASON_REFRESH_TARGET) //We'll repick our targets as there could be more better targets to attack
 	return ..()
 
-/datum/ai_behavior/carbon/xeno/deal_with_obstacle()
-	if(world.time < mob_parent.next_move)
-		return
-
-	var/list/things_nearby = range(mob_parent, 1) //Rather than doing multiple range() checks we can just archive it here for just this deal_with_obstacle
-	for(var/obj/structure/obstacle in things_nearby)
-		if(obstacle.resistance_flags & XENO_DAMAGEABLE)
-			var/mob/living/carbon/xenomorph/xeno = mob_parent
-			INVOKE_ASYNC(obstacle, /atom.proc/attack_alien, xeno)
-			mob_parent.face_atom(obstacle)
-			xeno.changeNext_move(xeno.xeno_caste.attack_delay)
-			return
-
-	//Cheat mode: insta open airlocks
-	for(var/obj/machinery/door/airlock/lock in things_nearby)
-		if(!lock.density) //Airlock is already open no need to force it open again
-			continue
-		if(lock.operating) //Airlock already doing something
-			continue
-		if(lock.welded) //It's welded, can't force that open
-			continue
-		lock.open(TRUE)
-		return //Don't try going on window frames after opening up airlocks dammit
-
-	//Teleport onto those window frames, we also can't attempt to attack said window frames so this isn't in the obstacles loop
-	for(var/obj/structure/window_frame/frame in things_nearby)
-		mob_parent.loc = frame.loc
-		return
+/datum/ai_behavior/carbon/xeno/deal_with_obstacle(datum/source, direction)
+	var/turf/obstacle_turf = get_step(mob_parent, direction)
+	for(var/thing in obstacle_turf.contents)
+		if(isstructure(thing))
+			if(istype(thing, /obj/structure/window_frame))
+				mob_parent.loc = obstacle_turf
+				return COMSIG_OBSTACLE_DEALT_WITH
+			if(istype(thing, /obj/structure/closet))
+				var/obj/structure/closet/closet = thing
+				if(closet.open(mob_parent))
+					return COMSIG_OBSTACLE_DEALT_WITH
+				return
+			var/obj/structure/obstacle = thing
+			if(obstacle.resistance_flags & XENO_DAMAGEABLE)
+				mob_parent.face_atom(obstacle)
+				INVOKE_ASYNC(src, .proc/attack_target, null, obstacle)
+				return COMSIG_OBSTACLE_DEALT_WITH
+		if(istype(thing, /obj/machinery/door/airlock))
+			var/obj/machinery/door/airlock/lock = thing
+			if(!lock.density) //Airlock is already open no need to force it open again
+				continue
+			if(lock.operating) //Airlock already doing something
+				continue
+			if(lock.welded) //It's welded, can't force that open
+				continue
+			lock.open(TRUE)
+			return COMSIG_OBSTACLE_DEALT_WITH
+	if(ISDIAGONALDIR(direction))
```

Comment:
if this is a diagonal move, we may have been blocked during our first diag move. So we look for obstacles on diagonals as well. This can lead to xenos being able to opens two doors at once, but this requires doors to be placed in a way that just doesn't happen in game, and this never happened in testing